### PR TITLE
sim: add a pseudo netdev backed by vpnkit

### PR DESCRIPTION
**Pull request by yamt**

### CODE REVIEW

File: arch/sim/src/sim/up_vpnkit.c [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+#include "vpnkit/protocol.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define ERROR(fmt, ...) \
+        fprintf(stderr, "up_vpnkit: " fmt "\n\r", ##__VA_ARGS__)
+#define INFO(fmt, ...) \
+        fprintf(stderr, "up_vpnkit: " fmt "\n\r", ##__VA_ARGS__)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_vpnkit.c]: \n\r to \r\n?

File: arch/sim/src/sim/vpnkit/protocol.c [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+#include <stdio.h>
+
+#include "protocol.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define ERROR(fmt, ...) \
+        fprintf(stderr, "up_vpnkit: " fmt "\n\r", ##__VA_ARGS__)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/vpnkit/protocol.c]: \n\r to \r\n?

File: arch/sim/src/sim/up_vpnkit.c [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+#define DEBUG(fmt, ...)
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+const char *g_vpnkit_socket_path = "/tmp/vpnkit-nuttx"; /* XXX config */
+struct vif_info g_vifinfo;
+int g_vpnkit_fd = -1;
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_vpnkit.c]: all can be static?

File: arch/sim/Kconfig [hunk 4]
```diff
@@ -117,15 +117,35 @@ config SIM_NETDEV
...
+	default SIM_NETDEV_TAP
+
+config SIM_NETDEV_TAP
+	bool "Simulated Network Device with TAP/WPCAP"
 	---help---
 		Build in support for a simulated network device using a TAP device on Linux or
 		WPCAP on Windows.
 
+config SIM_NETDEV_VPNKIT
+	bool "Simulated Network Device with VPNKit"
```
Comment (xiaoxiang781216) [inline arch/sim/Kconfig]: Could you update the related section in:
boards\sim\sim\sim\README.txt
boards\sim\sim\sim\NETWORK-LINUX.txt

File: arch/sim/src/sim/vpnkit/negotiate.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+#include <string.h>
+
+#include "protocol.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define ERROR(fmt, ...) \
+        fprintf(stderr, "up_vpnkit: " fmt "\n\r", ##__VA_ARGS__)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/vpnkit/negotiate.c]: change to \r\n

**Comment by xiaoxiang781216** [comment]
@yamt could you tell me which case should we prefer vpnkit than tap?

**Comment by yamt** [comment]
@xiaoxiang781216 
a few reasons to prefer vpnkit:
* works on macos
* works without root privilege

File: arch/sim/src/sim/vpnkit/negotiate.c (hunk 5 above)
Comment (yamt) [reply arch/sim/src/sim/vpnkit/negotiate.c]: done

File: arch/sim/Kconfig (hunk 4 above)
Comment (yamt) [reply arch/sim/Kconfig]: will do

File: arch/sim/src/sim/up_vpnkit.c (hunk 3 above)
Comment (yamt) [reply arch/sim/src/sim/up_vpnkit.c]: will do

File: arch/sim/src/sim/vpnkit/protocol.c (hunk 2 above)
Comment (yamt) [reply arch/sim/src/sim/vpnkit/protocol.c]: done

File: arch/sim/src/sim/up_vpnkit.c (hunk 1 above)
Comment (yamt) [reply arch/sim/src/sim/up_vpnkit.c]: done

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216
> a few reasons to prefer vpnkit:
> 
> * works on macos
> * works without root privilege

It's great if you can add the setup instruction.

File: arch/sim/Kconfig (hunk 4 above)
Comment (yamt) [reply arch/sim/Kconfig]: done

File: arch/sim/src/sim/up_vpnkit.c (hunk 3 above)
Comment (yamt) [reply arch/sim/src/sim/up_vpnkit.c]: done

**Pull request merged**